client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx

```diff
@@ -8,35 +8,44 @@ import { useSelector } from 'calypso/state';
 import { isUserLoggedIn } from 'calypso/state/current-user/selectors';
 import VideoPressIntroBackground from '../../steps-repository/intro/videopress-intro-background';
 import { useStepRouteTracking } from './hooks/use-step-route-tracking';
-import type { Flow, StepperStep } from '../../types';
+import type { Flow, Navigate, StepperStep } from '../../types';
 
 type StepRouteProps = {
 	step: StepperStep;
 	flow: Flow;
 	showWooLogo: boolean;
 	renderStep: ( step: StepperStep ) => JSX.Element | null;
+	navigate: Navigate< StepperStep[] >;
 };
 
-//TODO: Check we can move RenderStep function to here and remove the renderStep prop
-const StepRoute = ( { step, flow, showWooLogo, renderStep }: StepRouteProps ) => {
+// TODO: Check we can move RenderStep function to here and remove the renderStep prop
+const StepRoute = ( { step, flow, showWooLogo, renderStep, navigate }: StepRouteProps ) => {
 	const userIsLoggedIn = useSelector( isUserLoggedIn );
-	const loginUrl = useLoginUrlForFlow( { flow } );
 	const stepContent = renderStep( step );
-	const shouldRedirectToLogin = step.requiresLoggedInUser && ! userIsLoggedIn;
-	const shouldSkipRender = shouldRedirectToLogin || ! stepContent;
+	const loginUrl = useLoginUrlForFlow( { flow } );
+	const shouldAuthUser = step.requiresLoggedInUser && ! userIsLoggedIn;
+	const shouldSkipRender = shouldAuthUser || ! stepContent;
+	const skipTracking = shouldAuthUser || ! stepContent;
+
+	const useBuiltItInAuth = flow.__experimentalUseBuiltinAuth;
 
 	useStepRouteTracking( {
 		flowName: flow.name,
 		stepSlug: step.slug,
-		skipTracking: shouldSkipRender,
+		skipTracking,
 		flowVariantSlug: flow.variantSlug,
 	} );
 
 	useEffect( () => {
-		if ( shouldRedirectToLogin ) {
+		if ( shouldAuthUser && ! useBuiltItInAuth ) {
 			window.location.assign( loginUrl );
 		}
-	}, [ loginUrl, shouldRedirectToLogin ] );
+	}, [ loginUrl, shouldAuthUser, useBuiltItInAuth ] );
+
+	if ( useBuiltItInAuth && shouldAuthUser && ! userIsLoggedIn ) {
+		navigate( 'user', undefined, true );
```

Comment:
I think the code gets more readable if we pass something like
`const extraData = {}; or even undefined` instead of passing `undefined` directly as a parameter.

Furthermore, is the `return null` necessary here?

Reply:
> Furthermore, is the return null necessary here?

Yes. This return of this will be rendered.